routing/payment_lifecycle.go

```diff
@@ -319,27 +322,42 @@ lifecycle:
 }
 
 // checkTimeout checks whether the payment has reached its timeout.
-func (p *paymentLifecycle) checkTimeout() error {
-	select {
-	case <-p.timeoutChan:
-		log.Warnf("payment attempt not completed before timeout")
-
+func (p *paymentLifecycle) checkTimeoutAndContext(ctx context.Context) error {
+	failPayment := func(reason channeldb.FailureReason) error {
 		// By marking the payment failed, depending on whether it has
 		// inflight HTLCs or not, its status will now either be
 		// `StatusInflight` or `StatusFailed`. In either case, no more
 		// HTLCs will be attempted.
-		err := p.router.cfg.Control.FailPayment(
-			p.identifier, channeldb.FailureReasonTimeout,
-		)
+		err := p.router.cfg.Control.FailPayment(p.identifier, reason)
 		if err != nil {
 			return fmt.Errorf("FailPayment got %w", err)
 		}
 
+		return nil
+	}
+
+	select {
+	case <-p.timeoutChan:
+		log.Warnf("payment attempt not completed before timeout")
+		err := failPayment(channeldb.FailureReasonTimeout)
+		if err != nil {
+			return err
+		}
+
+	case <-ctx.Done():
```

Comment:
I'd argue, that if the context times out we could instead fail the payment with `FailureReasonTimeout`. Would need to check if the `ctx.Err()` is `context.DeadlineExceeded`.

Reply:
I'd like to get rid of `p.timeoutChan` by moving the timeout to the context (but only if `cancelable` is set on the payRequest - see @bitromortac's comment), however then we couldn't distinguish between a context cancellation and a timeout cancellation inside the payment loop.

Reply:
Thanks for pointing this out. I've removed the `p.timeoutChan` and replaced it with a deadline dependent context.